rules/scripts/hsm_init.template.sh

```diff
@@ -0,0 +1,184 @@
+    o)
+      FLAGS_OUT_TAR="${OPTARG}"
+      ;;
+    w)
+      FLAGS_WIPE=true
+      ;;
+    *)
+      usage
+      ;;
+  esac
+done
+shift $((OPTIND - 1))
+
+if [[ "$#" -gt 0 ]]; then
+  echo "Unexpected arguments:" "$@" >&2
+  exit 1
+fi
+
+if [[ ! -x "${HSMTOOL_BIN}" ]]; then
+  echo "Error: HSMTOOL_BIN is not set or not executable."
+  exit 1
+fi
+
+if [[ -n "${FLAGS_IN_TAR}" && "${FLAGS_IN_TAR}" != *.tar.gz  ]]; then
+  echo "Error: Input tarball must have .tar.gz extension."
+  exit 1
+fi
+
+if [[ -n "${FLAGS_OUT_TAR}" && "${FLAGS_OUT_TAR}" != *.tar.gz  ]]; then
+  echo "Error: Output tarball must have .tar.gz extension."
+  exit 1
+fi
+
+if [[ ${#CERTGEN_TEMPLATES[@]} -ne ${#CERTGEN_KEYS[@]} ]]; then
+  echo "Error: Number of certgen templates and keys do not match."
+  exit 1
+fi
+
+if [[ ${#CERTGEN_TEMPLATES[@]} -ne ${#CERTGEN_ENDORSING_KEYS[@]} ]]; then
+  echo "Error: Number of certgen templates and endorsing keys do not match."
+  exit 1
+fi
+
+if [[ -n "${FLAGS_IN_TAR}" ]]; then
+  if [[ ! -f "${FLAGS_IN_TAR}" ]]; then
+    echo "Error: Input tarball does not exist."
+    exit 1
+  fi
+  echo "Extracting input tarball ${FLAGS_IN_TAR}"
+  tar -xzf "${FLAGS_IN_TAR}"
+fi
+
+# certgen generates a certificate for the given config file and signs it with
+# the given CA key.
+certgen () {
+  config_basename="${1%.conf}"
+  ca_key="${2}"
+  endorsing_key="${3}"
+
+  # Generate the CA certificate for the root CA.
+  SOFTHSM2_CONF="${FLAGS_SOFTHSM_CONFIG}" \
+  PKCS11_MODULE_PATH="${FLAGS_HSMTOOL_MODULE}" \
+  openssl req -new -engine pkcs11 -keyform engine \
+    -config "${config_basename}.conf" \
+    -out "${OUTDIR_CA}/${config_basename}.csr" \
+    -key "pkcs11:pin-value=${HSMTOOL_PIN};object=${ca_key};token=${FLAGS_HSMTOOL_TOKEN}"
+
+  SOFTHSM2_CONF="${FLAGS_SOFTHSM_CONFIG}" \
+  PKCS11_MODULE_PATH="${FLAGS_HSMTOOL_MODULE}" \
+  openssl x509 -req -engine pkcs11 -keyform engine \
+    -in "${OUTDIR_CA}/${config_basename}.csr" \
+    -out "${OUTDIR_CA}/${config_basename}.pem" \
+    -days 3650 \
+    -extfile "${config_basename}.conf" \
+    -extensions v3_ca \
+    -signkey "pkcs11:pin-value=${HSMTOOL_PIN};object=${endorsing_key};token=${FLAGS_HSMTOOL_TOKEN}"
+
+  rm "${OUTDIR_CA}/${config_basename}.csr"
+}
+
+# Create output directory for HSM exported files.
+mkdir -p "${OUTDIR_HSM}" "${OUTDIR_CA}" "${OUTDIR_PUB}"
+
+echo "softhsm-config: ${FLAGS_SOFTHSM_CONFIG}"
+echo "hsmtool-module: ${FLAGS_HSMTOOL_MODULE}"
+echo "hsmtool-token: ${FLAGS_HSMTOOL_TOKEN}"
+echo "hsmtool-pin: ${FLAGS_HSMTOOL_PIN}"
+echo "hsmtool-bin: ${HSMTOOL_BIN}"
+
+if [[ "${FLAGS_WIPE}" == true ]]; then
+  SOFTHSM2_CONF="${FLAGS_SOFTHSM_CONFIG}"  \
+  HSMTOOL_MODULE="${FLAGS_HSMTOOL_MODULE}" \
+  HSMTOOL_USER="user"                      \
+  HSMTOOL_TOKEN="${FLAGS_HSMTOOL_TOKEN}"   \
+  HSMTOOL_PIN="${FLAGS_HSMTOOL_PIN}"       \
+  "${HSMTOOL_BIN}"                         \
+    --logging=info                         \
+    --token="${FLAGS_HSMTOOL_TOKEN}"       \
```

Comment:
Do you need `--token` when `HSMTOOL_TOKEN` is set?  (if you do, I think that's an hsmtool bug)

Reply:
Fixed. The flag was not needed.